R/nm-file.R

```diff
@@ -115,46 +109,47 @@ nm_data <- function(.mod, .sep = ",") {
 # PRIVATE HELPERS AND IMPLEMENTATION
 
 #' Implementation function for reading NONMEM files
-#' @importFrom readr read_table read_table2 cols read_lines
+#' @importFrom tibble as_tibble
+#' @importFrom data.table fread
 #' @importFrom stringr str_detect
 #' @keywords internal
-nm_file_impl <- function(.path, .est_method) {
+nm_file_impl <- function(.path) {
   # read file and find top of table
   verbose_msg(glue("Reading {basename(.path)}"))
-  .txt <- read_lines(.path)
-  .est <- which(str_detect(.txt, "^ *TABLE NO"))
-  if (length(.est) == 0) {
-    stop("Found no 'TABLE NO...' lines in file. Not a NONMEM output file.")
-  }
 
-  if (is.null(.est_method)) {
-    .est_method <- length(.est)
-  }
-  if (.est_method == "fail") {
-    if (length(.est) > 1) {
-      warning(glue("{.path} has {length(.est)} tables in it, but .est_method='fail' expects only one table per file."), call. = FALSE)
-      return(invisible(NULL))
+  # get line numbers of TABLE lines
+  checkmate::assert_file_exists(.path)
+
+  # read the file, but catch warning that tells us there are multiple tables
+  W <- NULL
+  .d <- withCallingHandlers({
+    as_tibble(fread(
+      .path,
+      na.strings = ".",
+      skip = 1,
+      verbose = FALSE
+    ))
+  },
+  warning = function(.w) {
+    if (str_detect(.w$message, "Stopped early.+TABLE NO")) {
+      W <<- paste(
+        "nm_file() does not support files with multiple tables in a single file.",
+        glue("{.path} appears to contain multiple tables and will be skipped."),
+        sep = "\n"
+      )
+      invokeRestart("muffleWarning")
     } else {
-      .est_method <- 1
+      warning(.w)
```

Comment:
Neat.  

I think you want the `invokeRestart` even if you go down the `else` arm.  Otherwise the warning is repeated.